codeintel: adding ctrl+p shortcut

Numbers88s opened the pull request:
Adds ctrl+p keyboard shortcut to search bar

<!-- Reminder: Have you updated the changelog and relevant docs (user docs, architecture diagram, etc) ? -->
<!-- Please notify @distribution if this PR contains changes to CI that may need to be cherry-picked on to patch release branches -->

olafurpg commented on client/web/src/search/input/SearchBox.tsx:
```diff
@@ -54,6 +55,7 @@ export interface SearchBoxProps
 }
 
 export const SearchBox: React.FunctionComponent<SearchBoxProps> = props => {
+    const [isFuzzyFinderVisible, setIsFuzzyFinderVisible] = useState(false)
```
We should guard `setIsFuzzyFinderVisible` so that it can only be updated in the repo view. Currently, you can make the fuzzy finder visible in non-repo pages, which has surprising behavior

- search for something
- trigger Cmd+P in the search page result, has no visible effect but calls `setIsFuzzyFinderVisible(true)`
- open a search result and the fuzzy finder becomes visible when loading the blob view

Numbers88s replied:
Oh good catch! Thank you. I fixed it.

olafurpg commented on client/web/src/components/fuzzyFinder/FuzzyFinder.tsx:
```diff
@@ -107,25 +101,30 @@ export interface Failed {
+async function downloadFilenamesGQL(
+    repository: string,
+    commit: string,
+    client: ApolloClient<object>
```
Very nice to get the apollo client working here 👍

Numbers88s replied:
I was in apollo mode 😄

olafurpg commented on client/web/src/components/fuzzyFinder/FuzzyFinder.tsx:
```diff
@@ -31,36 +28,33 @@ export interface FuzzyFinderProps {
+    const apolloClient = useApolloClient()
     const [fsm, setFsm] = useState<FuzzyFSM>({ key: 'empty' })
 
+    const { repoName = '', commitID = '' } = parseBrowserRepoURL(location.pathname + location.search + location.hash)
```
TIL, I didn't know you could provide defaults like this

olafurpg commented on client/web/src/components/fuzzyFinder/FuzzyFinder.tsx:
```diff
@@ -31,36 +28,33 @@ export interface FuzzyFinderProps {
 
+    const { repoName = '', commitID = '' } = parseBrowserRepoURL(location.pathname + location.search + location.hash)
+
+    if (repoName === 'search' || !repoName) {
```
I thought this could break in other non-repo pages like site-admin or batch changes but I wasn't able to reproduce problematic behavior. It might be worth confirming with somebody from the frontend-platform team what's the canonical way to identify a repo URL.

Numbers88s replied:
Yeah, I'll point it out to the frontend-platform, its kinda difficult since there is no specific route (_I think_)

valerybugakov replied:
The logic that we use to understand on which page are we is located here: https://github.com/sourcegraph/sourcegraph/blob/2b1e04574ad1360cf235d3c3b814d25ce1953b6a/client/web/src/Layout.tsx#L137-L142

The solution for the `FuzzyFinder` component would be:
1. Create another value: `isRepositoryRelatedPage` from existing value like `isSearchRelatedPage`, `isSearchRelatedPage`
2. Pass it down the tree, probably to `SearchNavbarItem` where we can enable/disable `<FuzzyFinder />` as @limitedmage suggested to avoid adding `FuzzyFinder` logic to the reusable `SearchBox` component.

olafurpg commented on client/web/src/components/fuzzyFinder/FuzzyModal.tsx:
```diff
@@ -77,27 +65,134 @@ export const FuzzyModal: React.FunctionComponent<FuzzyModalProps> = props => {
     // on a button at the bottom of the result list.
     const [maxResults, setMaxResults] = useState(props.initialMaxResults)
 
-    const state: FuzzyModalState = {
```
I'm glad to see you update my janky code to become more idiomatic React. I hadn't used react since 2014 when I wrote this PR and I'm not fully up-to-date on how to properly structure code 😅

Numbers88s replied:
Oh yeah, that was great! Hooks changed it a bit huh 👍

olafurpg replied:
Hooks are very cool and I need more exposure to them to fully grok how to use them effectively

olafurpg reviewed (changes requested):
I just tried this locally and it works great 🎉  I'm super excited to see this issue finally resolved. I have only one comment, otherwise I think this is ready for review by frontend-platform

olafurpg commented on client/web/src/components/fuzzyFinder/FuzzyFinder.tsx:
```diff
@@ -32,9 +32,14 @@ export const FuzzyFinder: React.FunctionComponent<FuzzyFinderProps> = props => {
     const { repoName = '', commitID = '' } = parseBrowserRepoURL(location.pathname + location.search + location.hash)
 
+    useEffect(() => {
+        if (repoName === 'search' || !repoName) {
```
Given this logic is likely to change, it would be good to refactor it into a method so that we don't have to search for patterns matching this condition.

olafurpg reviewed (commented):
I haven't tried the latest changed locally. Will do a more thorough review tomorrow!

limitedmage commented on client/web/src/search/input/SearchBox.tsx:
```diff
@@ -84,11 +87,27 @@ export const SearchBox: React.FunctionComponent<SearchBoxProps> = props => {
                 </div>
             </div>
             <SearchButton hideHelpButton={props.hideHelpButton} className={styles.searchBoxButton} />
+            {isFuzzyFinderVisible &&
+                !isErrorLike(props.settingsCascade.final) &&
+                !isErrorLike(props.settingsCascade.final?.experimentalFeatures) &&
+                props.settingsCascade.final?.experimentalFeatures?.fuzzyFinder && (
+                    <FuzzyFinder
+                        caseInsensitiveFileCountThreshold={
+                            props.settingsCascade.final?.experimentalFeatures
+                                ?.fuzzyFinderCaseInsensitiveFileCountThreshold
+                        }
+                        setIsVisible={bool => setIsFuzzyFinderVisible(bool)}
+                    />
+                )}
```
Not sure if this is the right place to have the fuzzy finder. We're looking into making the SearchBox a reusable component in a number of places, including the code monitor and code insights creation pages. So the fuzzy finder should be moved somewhere else so it's not multiplied. Maybe the global navbar or the Layout component?

Numbers88s replied:
We need it in the SearchBox so we can trigger it as part of the `Searchbox` being automatically focused on from the LazyMonacoQueryInput component.

limitedmage replied:
In that case, you could try to put it in whatever component renders the main search box instead (eg. the navbar or the homepage). `SearchBox.tsx`, going forward, is meant to be reusable.

Numbers88s replied:
Okay. Going forward the `SearchBox.tsx` is meant to be reusable.  We just need it in the repo view so we have it disabled in the homepage. Is the navbar component the one that is rendered in the repo view?

limitedmage replied:
Yes, this is probably the best place for it: https://sourcegraph.com/github.com/sourcegraph/sourcegraph/-/blob/client/web/src/search/input/SearchNavbarItem.tsx#L48:22 

Alternatively, why not do it in the file/repo view instead of the searchbox? Why are we using Monaco to detect events for things that don't affect the Monaco input itself?

Numbers88s replied:
Great! Yeah, if we don't put it in the Monaco Input when the SearchBar is focused on and you do the shortcut (ctrl+p) you get Chrome's Print dialog 🙃 .   I'll look at that component, thanks for link 👍

limitedmage replied:
Oh Monaco 🤦‍♀️

valerybugakov commented on client/web/src/search/input/SearchBox.tsx:
```diff
@@ -84,11 +87,27 @@ export const SearchBox: React.FunctionComponent<SearchBoxProps> = props => {
             </div>
             <SearchButton hideHelpButton={props.hideHelpButton} className={styles.searchBoxButton} />
+            {isFuzzyFinderVisible &&
+                !isErrorLike(props.settingsCascade.final) &&
+                !isErrorLike(props.settingsCascade.final?.experimentalFeatures) &&
+                props.settingsCascade.final?.experimentalFeatures?.fuzzyFinder && (
```
Maybe we can move @vovakulikov's `getExperimentalFeatures` helper to the shared utils folder and reuse it here to simplify this logic a bit?
https://github.com/sourcegraph/sourcegraph/blob/a68ae7530d09d5e000595681e48d7b6f2e38bc4f/client/web/src/insights/utils/get-experimental-features.ts#L6-L17

Numbers88s replied:
Is that being used somewhere? I see it declared but not used. I am happy to move it just don't if I am missing something

valerybugakov replied:
Correct, it's not used anywhere. Feel free to move it to a shared folder.

valerybugakov reviewed (approved):
I tried it locally, and it looks awesome!

One thing I noticed with the help menu, probably not related to this PR:

<img width="551" alt="Screenshot 2021-10-01 at 11 51 01" src="https://user-images.githubusercontent.com/3846380/135598429-f1a41fe8-6f50-4dd6-9f69-6f83a86b7836.png">

Maybe it's easy to fix this layout issue. 🙂

Numbers88s commented on client/web/src/components/fuzzyFinder/FuzzyFinder.tsx:
```diff
@@ -31,36 +28,38 @@ export interface FuzzyFinderProps {
-
     // The state machine of the fuzzy finder. See `FuzzyFSM` for more details
     // about the state transititions.
+    const apolloClient = useApolloClient()
```
Thank you so much for showing me this way. It's good to know as I continue to refactor to Apollo.

limitedmage reviewed (approved):
Looks good to me for the search box code 👍  (please have someone else review the FuzzyFinder and FuzzyModal)

olafurpg reviewed (changes requested):
This is getting really close 😁 Thank you everybody for the helpful reviews!

I just tried this locally and hit on three cases where I observed unexpected behavior:

- Pressing Cmd+P activates the browser's print dialog when I'm a blob page and the search bar is not focused. The Cmd+P shortcut works as expected (activates the fuzzy finder) when the search bar is focused.
- I'm still able to silently activate the fuzzy finder while on a search page. To reproduce
	1. Make sure fuzzy finder is not activated
	2. Open the URL https://sourcegraph.test/search?q=Sources+repo:sourcegraph/sourcegraph&patternType=literal
	3. Trigger Cmd+P, which has no visible effect because we're on a search page
	4. Click on a search result to open the blob page, the fuzzy finder modal loads because of step 3
- Pressing the browser's "back" button does not deactivate the fuzzy finder. This bug existed before this PR so we can deal with it separately, but it might be fixable together with the issue above where the fuzzy finder is silently activated on a search page.

Numbers88s commented:
Thanks @olafurpg and everyone for all the reviews!

I have fixed two of the three cases you outlined above @olafurpg:
- Cmd+P activates the fuzzy finder whether the search bar is focused or not.
- No longer can we silently activate the fuzzy finder while on a search page.

I'll have to look a little further on the third one, but we can circle back to it later after we merge this one. 

I've also simplified the fuzzy finder logic by moving the `getExperimentalFeatures` to a shared directory.

And fixed the layout issue with the shortcut modal.

Let me know what you think or if you find anything else. I've tested it as much as I can, but please let me know ✌️

olafurpg reviewed (approved):
LGTM 👍 Thank you @Numbers88s I can't wait to start using this change 😄

olafurpg commented:
> I'll have to look a little further on the third one, but we can circle back to it later after we merge this one.

This issue seems to be fixed with your latest changes. I'm not longer able to reproduce.
